presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,534 @@
+            String taskInstanceId,
+            TaskId taskId,
+            OutputBuffers outputBuffers,
+            StateMachine<BufferState> state,
+            TempStorage tempStorage,
+            DataSize threshold)
+            throws IllegalArgumentException
+    {
+        try {
+            requireNonNull(taskId, "taskId is null");
+            this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceIs is null");
+            this.outputBuffers = requireNonNull(outputBuffers, "outputBuffers is null");
+            this.state = requireNonNull(state, "state is null");
+            this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+            this.tempStorage.initialize();
+            this.threshold = threshold;
+
+            context = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == OutputBuffers.BufferType.SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public ListenableFuture<BufferResult> get(OutputBuffers.OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        requireNonNull(bufferId, "outputBufferId is null");
+        checkArgument(bufferId.getId() == 0, "Invalid buffer id");
+        checkArgument(maxSize.toBytes() > 0, "maxSize must be at least 1 byte");
+        acknowledge(bufferId, token);
```

Comment:
The `get` method acknowledges all the pages received from the last `get` request. `token` refers to the page we'd like to start reading from, so we're passing in `token` into acknowledge to remove all previous pages from the buffer. 

I believe other OutputBuffer implementations call `getPages` from the `ClientBuffer` class, which calls `acknowledgePages` to take care of this process. Since we're not dealing with a ClientBuffer here, I called `acknowledge` directly in my implementation.